### ISSUE TITLE
Where can I find the default template used by sp.modal function?

### ISSUE THREAD

**Issue by pflantzdog27** [service-portal/documentation#251]
Looking to edit the template being used to generate the default modal rendered by sp.modal function. I thought the template would be found in the sp_ng_template table -- but no luck. 

Any help would be appreciated!

**Comment by CapaJC** [comment]
The sp_modal.xml template is an XML file in the filesystem and not currently available in the database for viewing.

There may be some cool trick to viewing XML files that are in the filesystem, but I'm unaware of one. Someone else can chime in if they know of one.

**Comment by nathanfirth** [comment]
@pflantzdog27 I believe so_modal is just a wrapper on top of UI Bootstrap. Here is a tutorial on how to do modals without the Service Portal directive: https://serviceportal.io/modal-windows-service-portal/

**Comment by CapaJC** [comment]
Here's the current content of that file (I changed ext from xml to txt):

[sp_modal.txt](https://github.com/service-portal/documentation/files/1174652/sp_modal.txt)

**Comment by pflantzdog27** [comment]
Thanks @CapaJC and @nathanfirth !

@CapaJC that template is something I noticed too. I went ahead and customized that template a bit and added it to the sp_ng_template table.

Here's my issue though and why I was hoping I could edit the existing template:
1. I don't think you can use the templateUrl parameter and the spModal wrapper. When I try to use the spModal wrapper, the modal template rendered is the same as default (not my custom template). 

2. When I use $uibModal instead of spModal the custom template, as defined in the templateUrl parameter, IS rendered. The problem is - how do you then pass the widget and widgetInput parameters that are available when using spModal, but seem unavailable to $uibModal? 

What is that magic sauce that allows the spModal wrapper to pass options to the ng-template? If I could pass options through to the template using $uibModal, I would be in business... 

```
$scope.modalInstance = $uibModal.open({   
	widget: widgetName,
	widgetInput: { 
		sys_id: id
	},
        templateUrl : 'myCustomTemplate.html',
	size : size,
	buttons: [
		{label:'${Cancel}', cancel: true}
	]
});
```

Thanks guys!

**Comment by pmuramalla** [comment]
templateUrl is not a parameter you could set, I think you already figured
that by now.

You could extend the spModal directive and give your own template that is
defined in the ng_template. There are few ways to do this. I will post
something when I am back at my computer.

**Comment by pflantzdog27** [comment]
Thanks @pmuramalla , looking forward to reading what you've come up with. 

Finding where directives live in service portal has been a problem I run into fairly frequently. I know there's the angular provider table but that list is definitely not all inclusive. I'm hoping that through your explanation of  how to extend the spModal directive, I'll live in whole new world where I can access any directive I want in the portal :)

@CapaJC as a bit of an aside: is there a quick trick to know when something (a template in this instance) is part of the filesystem versus when it's available in the database for viewing? Should I just assume that if it's not where I thought it should be, it's probably part of filesystem?

Thanks!

**Comment by pmuramalla** [comment]
I thought creating a document in the community helps us to get some
feedback about the approach. I apologize for keep you waiting, it took
longer than I expected. It is about sp-model but you could use the same
approach for sp-modal

https://community.servicenow.com/docs/DOC-7450

**Comment by pflantzdog27** [comment]
Great work @pmuramalla - you really went above and beyond on this one. Appreciate all your efforts. I'm going through it now and will report back my findings

**Comment by pflantzdog27** [comment]
@pmuramalla I use spModal from the client script, not as a tag or element. spModel is a directive, but I'm not sure spModal is also a directive. Believe it's an api service, no? That difference makes the documentation hard to use.

**Comment by pmuramalla** [comment]
you are right.. I totally missed it, did learn some thing new. the pattern
I specified in the document wouldn't directly help extend spModal.

see if you can use the same pattern to extend the spModal factory ? I will
try that if I get some time this weekend.

**Comment by pflantzdog27** [comment]
None the less - knowing how to edit directives like spModel is hugely valuable, in and of itself, so thanks for sharing!

**Comment by pmuramalla** [comment]
I couldn't extend spModal in a way I would like to extend, which is not to
change / modify / reimplement base spModal . I end up creating an
angular-service-factory, custspModal using spModal. but change the code so
that we can pass optional variable templateUrl. As I am not extending
spModal - you will not get any latest updates servicenow has implemented
for spModal factory. Further these updates may not have to come as part of
an sn upgrade.

Let me know if you need update set to try out.

**Comment by dylanlindgren** [comment]
Note that none of the above is necessary in Jakarta, as in this release the `spModel` directive actually accepts a parameter called `template-url`. You can simply create a `sp_ng_template` record on the widget with your own custom template, and supply the name of that template through this parameter and it will be used instead of the default template (`sp-model.xml`).

This is an undocumented and I'm sure unsupported feature in Jakarta, however if you're willing to extend the `spModel` directive as has been discussed in this thread then using this parameter is a more light-weight alternative to that.
